### PR TITLE
implement liquidity incentive for dex

### LINKED ISSUE #175: Liquidity incentive for DEX
https://gist.github.com/xlc/983451eaaba9cd0cd883a3dd45a4d359

### PR DESCRIPTION
**Pull request by zjb0807**
Closes #175

### CODE REVIEW

File: modules/dex/src/lib.rs [hunk 1]
```diff
@@ -23,7 +24,15 @@ type CurrencyIdOf<T> = <<T as Trait>::Currency as MultiCurrency<<T as system::Tr
 pub trait Trait: system::Trait {
 	type Event: From<Event<Self>> + Into<<Self as system::Trait>::Event>;
 	type Currency: MultiCurrencyExtended<Self::AccountId>;
-	type Share: Parameter + Member + AtLeast32Bit + Default + Copy + MaybeSerializeDeserialize;
+	type Share: Into<BalanceOf<Self>>
+		+ From<BalanceOf<Self>>
```
Comment (xlc) [inline modules/dex/src/lib.rs:28]: `Into<BalanceOf<Self>> + From<BalanceOf<Self>>` basically means it have to be same type of `BalanceOf<Self>`

Do you need both? if not, remove the not used one. if yes, then just `type Share<T> = BalanceOf<T>` in global scope to keep thing simpler.

File: modules/dex/src/lib.rs [hunk 2]
```diff
@@ -56,9 +66,14 @@ decl_error! {
...
 decl_storage! {
 	trait Store for Module<T: Trait> as Dex {
+		LiquidityIncentiveRate get(fn liquidity_incentive_rate) config(): Rate;
 		LiquidityPool get(fn liquidity_pool): map hasher(twox_64_concat) CurrencyIdOf<T> => (BalanceOf<T>, BalanceOf<T>);
 		TotalShares get(fn total_shares): map hasher(twox_64_concat) CurrencyIdOf<T> => T::Share;
 		Shares get(fn shares): double_map hasher(twox_64_concat) CurrencyIdOf<T>, hasher(twox_64_concat) T::AccountId => T::Share;
+		TotalDebits get(fn total_debits): map hasher(twox_64_concat) CurrencyIdOf<T> => T::Share;
+		Debits get(fn debits): double_map hasher(twox_64_concat) CurrencyIdOf<T>, hasher(twox_64_concat) T::AccountId => T::Share;
+		TotalInterest get(fn total_interest): map hasher(twox_64_concat) CurrencyIdOf<T> => T::Share;
+		Withdrawn get(fn withdrawn): double_map hasher(twox_64_concat) CurrencyIdOf<T>, hasher(twox_64_concat) T::AccountId => T::Share;
```
Comment (xlc) [inline modules/dex/src/lib.rs:76]: this is for PoC purpose, no need to have this storage

File: modules/dex/src/lib.rs [hunk 3]
```diff
@@ -490,6 +520,98 @@ impl<T: Trait> Module<T> {
...
+		let proportion = Ratio::from_rational(shares, Self::total_shares(currency_id));
+		let total_interest = Self::total_interest(currency_id);
+		let withdrawn_interest = proportion.saturating_mul_int(&total_interest).saturating_sub(debits);
+		<Debits<T>>::mutate(currency_id, &who, |debits| {
+			*debits = debits.saturating_add(withdrawn_interest);
+		});
+		<TotalDebits<T>>::mutate(currency_id, |debits| {
+			*debits = debits.saturating_add(withdrawn_interest);
+		});
+		<Withdrawn<T>>::mutate(currency_id, &who, |withdrawn| {
```
Comment (xlc) [inline modules/dex/src/lib.rs:581]: should use `T::Currency::deposit` to mint aUSD for user

Comment (wangjj9219) [reply modules/dex/src/lib.rs:581]: deposit aUSD without cdp backed, must increase debit to cdp treasury at the same time.
Extend trait `CDPTreasury`, add `deposit_unbacked_debit` function to do it.

File: modules/dex/src/lib.rs (hunk 1 above)
Comment (zjb0807) [reply modules/dex/src/lib.rs:28]: `type Share<T> = BalanceOf<T>` the way will not use the associated type `Share`

File: runtime/tests/integration_test.rs [hunk 4]
```diff
@@ -152,6 +152,9 @@ mod tests {
 				assert_eq!(DexModule::liquidity_pool(XBTC), (10002, 10002000));
 				assert_ok!(DexModule::add_liquidity(origin_of(AccountId::from(BOB)), XBTC, 1, 1001));
 				assert_eq!(DexModule::liquidity_pool(XBTC), (10003, 10003000));
+				DexModule::on_initialize(0);
+				assert_eq!(DexModule::total_shares(XBTC), 10002998);
```
[the remark this answers is not in the document]
Comment (zjb0807) [reply runtime/tests/integration_test.rs:156]: Accumulate interest not work here.

**Pull request merged**